File: src/ts/src/type_def.ts
/**
 * 三维坐标，单位米，格式 [x, y, z]
 */
export type XYZ = [number, number, number];

/**
 * 链路可达性检查的可选参数。
 */
export interface RadioCheckOptions {
    /** 信号频率（MHz），默认 2400 */
    frequencyMHz?: number;
    /** 发射功率（dBm），默认 20 */
    txPowerDbm?: number;
    /** 接收灵敏度（dBm），默认 -90 */
    rxSensitivityDbm?: number;
    /** 菲涅尔区清晰度阈值（0~1），默认 0.6 */
    fresnelZoneRatio?: number;
    /** 是否跳过菲涅尔区体积检测（默认 false） */
    skipFresnelZoneCheck?: boolean;
    /** 是否启用多径反射计算（默认 false） */
    enableMultipath?: boolean;
    /** 最大反射路径数（默认 8） */
    maxReflectionPaths?: number;
    /** 反射路径最大长度与直射距离的比值（默认 2.0） */
    maxReflectionPathLengthRatio?: number;
    /** 发射端默认天线增益（dBi），默认 0 */
    defaultTxAntennaGain_dBi?: number;
    /** 接收端默认天线增益（dBi），默认 0 */
    defaultRxAntennaGain_dBi?: number;
    /** 是否启用天线方向图增益修正（默认 false） */
    enableAntennaPattern?: boolean;
    /** 是否启用多天线互耦合修正（默认 false） */
    enableMutualCoupling?: boolean;
    /** 互耦合可忽略的天线间距阈值（单位：波长数），默认 10 */
    couplingNegligibleThresholdWavelengths?: number;
    /** 是否启用同频 SINR（信噪干扰比）计算（默认 false） */
    enableSINR?: number;
    /** 接收机噪声带宽（Hz），默认 20e6 */
    receiverBandwidthHz?: number;
    /** SINR 最低阈值（dB），默认 10 */
    minSINR_dB?: number;
    /** 是否启用频段隔离计算（默认 false） */
    enableFrequencyIsolation?: boolean;
    /** 是否启用近场修正（默认 false） */
    enableNearFieldCorrection?: boolean;
    /** 保留字段，暂未使用 */
    enableNodeBodyOcclusion?: boolean;
}

/**
 * 链路可达性检查请求消息
 */
export interface CheckReachabilityRequest {
    aTx: XYZ;
    bRx: XYZ;
    options?: RadioCheckOptions;
}

/**
 * 更新对象位置的消息
 */
export interface UpdateObjectPosRequest {
    objectId: string;
    position: XYZ;
}

/**
 * 更新对象的电磁属性
 */
export interface UpdateMeshRadioMaterialRequest {
    meshId: string;
    materialId?: string;
    thickness_m?: number;
}

/**
 * 电磁材料属性定义
 */
export interface RadioMaterialProperties {
    id: string;
    displayName: string;
    penetrationLoss_dBPerMeter: number;
    reflectionCoefficient: number;
    defaultThickness_m: number;
}

/**
 * 摇杆输入消息
 */
export interface JoyStickInput {
    vx?: number;
    vy?: number;
    vz?: number;
    yawRate?: number;
}

File: src/rb/PhantasyIslandRubyRemoteControl/radio/type_def.rb
# Ruby 等价类型定义，对应 TypeScript TypeBox 定义的类型。

# XYZ — 三维坐标，单位米，格式 [x, y, z]
# 在 Ruby 中直接使用数组表示即可

class RadioCheckOptions
  attr_accessor :frequencyMHz, :txPowerDbm, :rxSensitivityDbm, :fresnelZoneRatio,
                :skipFresnelZoneCheck, :enableMultipath, :maxReflectionPaths,
                :maxReflectionPathLengthRatio, :defaultTxAntennaGain_dBi,
                :defaultRxAntennaGain_dBi, :enableAntennaPattern, :enableMutualCoupling,
                :couplingNegligibleThresholdWavelengths, :enableSINR, :receiverBandwidthHz,
                :minSINR_dB, :enableFrequencyIsolation, :enableNearFieldCorrection,
                :enableNodeBodyOcclusion

  def initialize(options = {})
    options.each { |k, v| send("#{k}=", v) if respond_to?("#{k}=") }
  end

  def to_h
    hash = {}
    instance_variables.each do |var|
      val = instance_variable_get(var)
      hash[var.to_s.delete('@')] = val unless val.nil?
    end
    hash
  end

  def self.from_h(data)
    new(data)
  end
end

class CheckReachabilityRequest
  attr_accessor :aTx, :bRx, :options

  def initialize(aTx:, bRx:, options: nil)
    @aTx = aTx
    @bRx = bRx
    @options = options
  end

  def to_h
    h = {
      'aTx' => @aTx,
      'bRx' => @bRx
    }
    h['options'] = @options.to_h if @options
    h
  end

  def self.from_h(data)
    opts = data['options'] ? RadioCheckOptions.from_h(data['options']) : nil
    new(
      aTx: data['aTx'],
      bRx: data['bRx'],
      options: opts
    )
  end
end

class UpdateObjectPosRequest
  attr_accessor :objectId, :position

  def initialize(objectId:, position:)
    @objectId = objectId
    @position = position
  end

  def to_h
    {
      'objectId' => @objectId,
      'position' => @position
    }
  end

  def self.from_h(data)
    new(
      objectId: data['objectId'],
      position: data['position']
    )
  end
end

class UpdateMeshRadioMaterialRequest
  attr_accessor :meshId, :materialId, :thickness_m

  def initialize(meshId:, materialId: nil, thickness_m: nil)
    @meshId = meshId
    @materialId = materialId
    @thickness_m = thickness_m
  end

  def to_h
    h = { 'meshId' => @meshId }
    h['materialId'] = @materialId unless @materialId.nil?
    h['thickness_m'] = @thickness_m unless @thickness_m.nil?
    h
  end

  def self.from_h(data)
    new(
      meshId: data['meshId'],
      materialId: data['materialId'],
      thickness_m: data['thickness_m']
    )
  end
end

class RadioMaterialProperties
  attr_accessor :id, :displayName, :penetrationLoss_dBPerMeter, :reflectionCoefficient, :defaultThickness_m

  def initialize(id:, displayName:, penetrationLoss_dBPerMeter:, reflectionCoefficient:, defaultThickness_m:)
    @id = id
    @displayName = displayName
    @penetrationLoss_dBPerMeter = penetrationLoss_dBPerMeter
    @reflectionCoefficient = reflectionCoefficient
    @defaultThickness_m = defaultThickness_m
  end

  def to_h
    {
      'id' => @id,
      'displayName' => @displayName,
      'penetrationLoss_dBPerMeter' => @penetrationLoss_dBPerMeter,
      'reflectionCoefficient' => @reflectionCoefficient,
      'defaultThickness_m' => @defaultThickness_m
    }
  end

  def self.from_h(data)
    new(
      id: data['id'],
      displayName: data['displayName'],
      penetrationLoss_dBPerMeter: data['penetrationLoss_dBPerMeter'],
      reflectionCoefficient: data['reflectionCoefficient'],
      defaultThickness_m: data['defaultThickness_m']
    )
  end
end

class JoyStickInput
  attr_accessor :vx, :vy, :vz, :yawRate

  def initialize(vx: 0.0, vy: 0.0, vz: 0.0, yawRate: 0.0)
    @vx = vx
    @vy = vy
    @vz = vz
    @yawRate = yawRate
  end

  def to_h
    {
      'vx' => @vx,
      'vy' => @vy,
      'vz' => @vz,
      'yawRate' => @yawRate
    }
  end

  def self.from_h(data)
    new(
      vx: data.fetch('vx', 0.0),
      vy: data.fetch('vy', 0.0),
      vz: data.fetch('vz', 0.0),
      yawRate: data.fetch('yawRate', 0.0)
    )
  end
end
